### PR TITLE
prov/net: RX operations now support io_uring

### PR DESCRIPTION
**Pull request by sydidelot**
This patch adds support of io_uring to RX operations.

The implementation is still incomplete because we need a mechanism
to cancel all pending io_uring operations and wait for them to
complete before we can close the ring. Pending receive operations
would cause the code to fail an assertion failure when the endpoint
is shut down.

### CODE REVIEW

**Comment by sydidelot** [comment]
@shefty It seems that I don't have the permission to re-open the older PR #8262, so I created this new PR instead.
Most of the code is identical to the older PR, except some minor changes to the `async_prefetch`.

The next step will be the implementation of a mechanism to cancel outstanding io_uring ops when the EP is abruptly shut down.
I will need your help for this task: `xnet_ep_flush_all_queues()` is assumed to be synchronous, which is no longer true with io_uring and I don't know what's the best solution for this problem.

**Comment by shefty** [comment]
Intel CI failures appear to be related:
```
- name:   fi_rdm_tagged_peek -p "net"
00:57:13    timestamp: 20230113-085707+0000
00:57:13    result: Fail

- name:   fi_rdm_tagged_pingpong -v -p "net"
01:03:55    timestamp: 20230113-090348+0000
01:03:55    result: Fail

- name:   fi_rdm_tagged_bw -v -U -p "net"
01:09:09    timestamp: 20230113-090904+0000
01:09:09    result: Fail

- name:   fi_multinode -C msg -n 3 -p "net"
01:14:30    timestamp: 20230113-091425+0000
01:14:30    result: Fail

- name:   fi_multinode -C rma -n 3 -p "net"
01:19:37    timestamp: 20230113-091926+0000
01:19:37    result: Fail
```
I restarted the other CI failures -- those looked like a connectivity problem.

**Comment by sydidelot** [comment]
Thanks. I'll take a look at the failing tests!

File: prov/net/src/xnet_progress.c [hunk 1]
```diff
@@ -902,6 +793,53 @@ static int xnet_recv_hdr(struct xnet_ep *ep)
...
+{
+	struct xnet_xfer_entry *rx_entry;
+	struct util_cq *cq;
+
+	rx_entry = ep->cur_rx.entry;
+	if (!rx_entry) {
+		if (ret)
+			goto disable_ep;
+		return;
+	}
```
[the remark this answers is not in the document]
Comment (sydidelot) [reply prov/net/src/xnet_progress.c:819]: There are some code paths where `ep->cur_rx.handler(ep);` returns 0 and `ep->cur_rx.entry` is NULL. This is for example the case where `ep->cur_rx.handler` is set to `xnet_op_read_req()`. This latter function calls `xnet_reset_rx()` just before returning 0.
If you prefer I can remove the check from `xnet_complete_rx()` and move it to the caller. For example:
```
if (ep->cur_rx.entry)                   
        xnet_complete_rx(ep, ret);      
else if (ret)                           
        xnet_ep_disable(ep, 0, NULL, 0);
```

**Comment by sydidelot** [comment]
FYI, I found the bug for the failing tests. It was introduced when I resolved a couple of conflicts after rebasing.

**Comment by shefty** [comment]
thanks!

**Pull request merged**